### PR TITLE
Better deletion tracking in ChangeSet

### PR DESCRIPTION
**Pull request by dcherian**
Now track both Path and NodeId rather than just Path. This helps us handle tricky cases where a committed node is updated, deleted, and created again in the same session.

### CODE REVIEW

File: icechunk/src/change_set.rs [hunk 1]
```diff
@@ -151,14 +141,24 @@ impl ChangeSet {
...
-        self.deleted_groups.contains(path)
-            || self.deleted_arrays.contains(path)
-            || path.ancestors().skip(1).any(|parent| self.is_deleted(&parent))
+    pub fn is_deleted(&self, path: &Path, node_id: &NodeId) -> bool {
+        dbg!(format!("checking is_deleted for {0} & {1}", path, node_id));
+        let key = (path.clone(), node_id.clone());
+        self.deleted_groups.contains(&key)
+            || self.deleted_arrays.contains(&key)
+            || path.ancestors().skip(1).any(|parent| {
+                self.get_group(&parent).is_some_and(|parent_node_id| {
```
Comment (dcherian) [inline icechunk/src/change_set.rs:154]: This is wrong, and an existing test fails. 

We are in `ChangeSet` and somehow need to know the `NodeId` for a node that is in the snapshot. How do I get that?

File: icechunk-python/tests/test_store.py [hunk 2]
```diff
@@ -0,0 +1,15 @@
+import icechunk
+import zarr
+
+
+async def test_store_clear() -> None:
```
Comment (dcherian) [inline icechunk-python/tests/test_store.py:5]: hooray hypothesis!

File: icechunk/src/change_set.rs (hunk 1 above)
Comment (paraseba) [reply icechunk/src/change_set.rs:154]: :thinking:  ... I think the problem is  with the `get_group` call right? That's what doesn't work... but it feels this is fixable? We need to know if a given path was deleted _in this changeset_, we shouldn't need the repo for that.

Comment (dcherian) [reply icechunk/src/change_set.rs:154]: We need to know if a *node* was deleted, not just the path (that's the issue :) )

File: icechunk/src/conflicts/basic_solver.rs [hunk 3]
```diff
@@ -90,10 +90,10 @@ impl BasicConflictSolver {
                     ChunkDoubleUpdate{..} if self.on_chunk_conflict == VersionSelection::Fail
                 ) ||
                 matches!(conflict,
-                    DeleteOfUpdatedArray(_) if self.fail_on_delete_of_updated_array
+                         DeleteOfUpdatedArray{..} if self.fail_on_delete_of_updated_array
```
Comment (dcherian) [inline icechunk/src/conflicts/basic_solver.rs:93]:
```suggestion
                     DeleteOfUpdatedArray{..} if self.fail_on_delete_of_updated_array
```

File: icechunk/src/conflicts/detector.rs [hunk 4]
```diff
@@ -169,7 +169,8 @@ impl ConflictSolver for ConflictDetector {
         let deletes_of_updated_arrays = stream::iter(
             current_changes.deleted_arrays().map(Ok),
         )
-        .try_filter_map(|path| async {
+        .try_filter_map(|(path, _node_id)| async {
+            // FIXME: reuse node_id?
```
Comment (dcherian) [inline icechunk/src/conflicts/detector.rs:173]:
```suggestion
```

File: icechunk/src/conflicts/basic_solver.rs [hunk 5]
```diff
@@ -90,10 +90,10 @@ impl BasicConflictSolver {
                     ChunkDoubleUpdate{..} if self.on_chunk_conflict == VersionSelection::Fail
                 ) ||
                 matches!(conflict,
-                    DeleteOfUpdatedArray(_) if self.fail_on_delete_of_updated_array
+                     DeleteOfUpdatedArray{..} if self.fail_on_delete_of_updated_array
```
Comment (dcherian) [inline icechunk/src/conflicts/basic_solver.rs:93]:
```suggestion
                    DeleteOfUpdatedArray{..} if self.fail_on_delete_of_updated_array
```

File: icechunk-python/tests/test_store.py (hunk 2 above)
Comment (paraseba) [reply icechunk-python/tests/test_store.py:5]: hooray!

File: icechunk/src/change_set.rs [hunk 6]
```diff
@@ -151,14 +141,13 @@ impl ChangeSet {
...
+            self.deleted_arrays.insert((path, node_id.clone()));
         }
     }
 
-    pub fn is_deleted(&self, path: &Path) -> bool {
-        self.deleted_groups.contains(path)
-            || self.deleted_arrays.contains(path)
-            || path.ancestors().skip(1).any(|parent| self.is_deleted(&parent))
+    pub fn is_deleted(&self, path: &Path, node_id: &NodeId) -> bool {
+        let key = (path.clone(), node_id.clone());
```
Comment (paraseba) [inline icechunk/src/change_set.rs:149]: I hate that Rust forces you to clone here.

File: icechunk/src/repository.rs [hunk 7]
```diff
@@ -355,8 +355,20 @@ impl Repository {
     /// Deletes of non existing groups will succeed.
     pub async fn delete_group(&mut self, path: Path) -> RepositoryResult<()> {
         match self.get_group(&path).await {
-            Ok(node) => {
-                self.change_set.delete_group(node.path, &node.id);
+            Ok(parent) => {
+                let nodes_iter: Vec<NodeSnapshot> = self.list_nodes().await?.collect();
```
Comment (paraseba) [inline icechunk/src/repository.rs:359]: How about calling `filter` on the result of `list_nodes` passing `is_prefix_match`. That way, you can construct a much smaller `Vec`

File: icechunk/src/repository.rs [hunk 8]
```diff
@@ -597,15 +609,8 @@ impl Repository {
...
-            self.list_nodes().await?.map(|node| (node.node_type(), node.path)).collect();
-
-        for (t, p) in to_delete {
-            match t {
-                NodeType::Group => self.delete_group(p).await?,
-                NodeType::Array => self.delete_array(p).await?,
-            }
-        }
+        #[allow(clippy::expect_used)]
+        self.delete_group(Path::try_from("/").expect("should not fail")).await?;
```
Comment (paraseba) [inline icechunk/src/repository.rs:613]: Is this too clever? I think the root group doesn't need to exist, for example they can have a root array.

File: icechunk/src/repository.rs [hunk 9]
```diff
@@ -931,6 +952,7 @@ pub async fn get_chunk(
     }
 }
 
+// Yields nodes in the base snapshot, applying any relevant updates in the changeset
```
Comment (paraseba) [inline icechunk/src/repository.rs:955]: add another / so it shows up in rendered documentation.

File: icechunk/src/repository.rs [hunk 10]
```diff
@@ -909,6 +914,22 @@ impl From<Repository> for ChangeSet {
     }
 }
 
+pub fn is_prefix_match(key: &str, prefix: &str) -> bool {
+    let tomatch =
+        if prefix != &String::from('/') { key.strip_prefix(prefix) } else { Some(key) };
```
Comment (paraseba) [inline icechunk/src/repository.rs:919]: Can you avoid the allocation by saying `prefix != "/"` ?

Comment (paraseba) [reply icechunk/src/repository.rs:919]: This code is tricky and somewhat hard to follow. I wonder, can we use `Path` methods for this? Path's should be able to tell you if they are parent/child right?

Comment (paraseba) [reply icechunk/src/repository.rs:919]: That would also skip the somewhat expensive conversion from Path to String

File: icechunk/src/repository.rs [hunk 11]
```diff
@@ -954,6 +976,8 @@ async fn updated_existing_nodes<'a>(
     Ok(updated_nodes)
 }
 
+// Yields nodes with the snapshot, applying any relevant updates in the changeset,
+// *and* new nodes in the changeset
```
Comment (paraseba) [inline icechunk/src/repository.rs:980]: see above, add a "/"

**Review by paraseba — APPROVED** [review]
This is great Deepak
